### PR TITLE
Fix Docker Engine RPC buffer overflows.

### LINKED ISSUE #812: The `repoDigest` output sometimes does not update on builds
### What happened?

When building a new Image, and logs show that the image was built and pushed to the remote registry, the `repoDigest` output sometimes does not change. This occurs only when the image name and tag (`imageName`) is  consistent between builds, suggesting either a race condition in either our provider or the Docker engine.

### Example

Build an image with the same `imageName` repeatedly, and an output printing the `repoDigest`. The `repoDigest` rarely will not show an update.

### Output of `pulumi about`

Latest Docker & Pulumi as of 10/25/2023

### Additional context

The relevant section of the build implementation:
1. Runs the `ImagePush` RPC against the Docker Engine
2. Streams the output
3. Runs the `InspectImageWithRaw` RPC against the Docker Engine, retrieving the image with the `imageName` property

By line 390 below, I would expect the engine's image store to have updated, and the `imageName` should refer to the newly built image. It appears that sometimes the result of the call returns the _previously_ built image.

https://github.com/pulumi/pulumi-docker/blob/902ce25e0df6a3c921629887e2f4015a3c0201d8/provider/image.go#L361-L391

### Contributing

Vote on this issue by adding a 👍 reaction. 
To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already).

### PR DESCRIPTION
**Pull request by AaronFriel**
Fix Docker Engine RPC buffer overflows.

Fixes #812, #778, #573.

In issue [#573](https://github.com/pulumi/pulumi-docker/issues/573), users would sometimes report the image push fails with an error "the image does not exist locally with the tag". This is likely due to Docker Engine RPCs overflowing the default `bufio.Scanner` maximum buffer size of 64k. Replacing the scanner with a `json.Decoder` enables us to handle large responses. When the image build fails, users should now see a warning or error before the provider pushes the built image.

In issue [#812](https://github.com/pulumi/pulumi-docker/issues/812), users would see the update operation on the Image resource succeed, but the output `repoDigest` would not change. In this PR, if for any reason the image fails to build, or update the Docker Engine's image store, push, or we detect inconsistency between the image ID returned by the `ImageBuild` RPC and what we expected, we fail the resource operation with an error.

In issue [#778](https://github.com/pulumi/pulumi-docker/issues/778), a repo digest would be chosen that doesn't match the name (sans tag) of the image pushed. As part of the additional checks, we fix this and ensure the normalized name of the repo digest matches the input image name. 

Additionally, in case there is a data race between `ImageBuild` and `ImagePush`, we extract the expected image IDs and repo digests from those operations and compare to the image store. As this is a net-new check and we could not reproduce a repo digest mismatch, this is a warning.

In a follow-up (#846) we may consider removing some of the defensive programming in this change.

### CODE REVIEW

File: provider/image.go [hunk 1]
```diff
@@ -401,13 +402,140 @@ func (p *dockerNativeProvider) dockerBuild(ctx context.Context,
...
+		// Compare on name, sans tag:
+		if imageName.Name() == ref.Name() {
+			repoDigest = ref
+			break
+		}
+	}
+	return repoDigest, err
+}
+
+// runImageBuild is a helper function that runs the image build and ensures that the correct image
```
Comment (guineveresaenger) [inline provider/image.go:438]:
```suggestion
// runImageBuild runs the image build and ensures that the correct image
```

File: provider/image.go [hunk 2]
```diff
@@ -673,24 +794,59 @@ func getRegistryAddrFromImage(imgName string) (string, error) {
...
+		err := decoder.Decode(&jm)
+		if err != nil {
+			if err == io.EOF {
+				break
+			}
+			return fmt.Errorf("error parsing Docker output: %v", err)
+		}
+		newOffset := decoder.InputOffset()
+		if newOffset-currentOffset > 64*1024 {
+			_ = p.host.Log(ctx, "info", urn, fmt.Sprintf("Read %d KiB from Docker", (newOffset-currentOffset)/1024))
```
Comment (guineveresaenger) [inline provider/image.go:814]: Would we like to remove this, as it is mostly for internal diagnostics? I'm happy also to keep for now but remove later.

File: provider/image.go [hunk 3]
```diff
@@ -401,13 +402,140 @@ func (p *dockerNativeProvider) dockerBuild(ctx context.Context,
...
+				_ = p.host.Log(ctx, "warning", urn,
+					fmt.Sprintf("Found in store does not match built image, expected label %s=%s, found %s",
+						buildIDLabel, buildIDValue, storedImage.Labels[buildIDLabel]))
+				return false, nil
+			}
+		}
+		return false, nil
+	}
+
+	for i := 0; i < 60; i++ {
```
Comment (guineveresaenger) [inline provider/image.go:520]: if I read this correctly, we are ensuring no race conditions between local image store and build results. Why 60 times? Have we seen a latency of > 5 seconds? I _think_ `docker.ImageList` is a fairly cheap call, so this is probably fine.

However we should catch the case where we try this for a minute straight with no results and notify of a timeout while polling.

File: provider/image.go [hunk 4]
```diff
@@ -401,13 +402,140 @@ func (p *dockerNativeProvider) dockerBuild(ctx context.Context,
...
+		if err != nil {
+			return "", err
+		}
+		if found {
+			break
+		}
+		time.Sleep(1 * time.Second)
+	}
+
+	_ = p.host.LogStatus(ctx, "info", urn, fmt.Sprintf("Image built successfully, local id %q", imageID))
```
Comment (guineveresaenger) [inline provider/image.go:531]: Should this persist to final log output? Can go either way here.
```suggestion
	_ = p.host.Log(ctx, "info", urn, fmt.Sprintf("Image built successfully, local id %q", imageID))
```

File: provider/image.go [hunk 5]
```diff
@@ -401,13 +402,140 @@ func (p *dockerNativeProvider) dockerBuild(ctx context.Context,
...
+				if storedImage.Labels[buildIDLabel] == buildIDValue {
+					return true, nil
+				}
+				_ = p.host.Log(ctx, "warning", urn,
+					fmt.Sprintf("Found in store does not match built image, expected label %s=%s, found %s",
+						buildIDLabel, buildIDValue, storedImage.Labels[buildIDLabel]))
+				return false, nil
+			}
+		}
+		return false, nil
```
Comment (guineveresaenger) [inline provider/image.go:517]: It may be worth logging the "not found" case? if the image isn't found even by image ID, this is surely a problem!

**Review by guineveresaenger — APPROVED** [review]
Largely this LGTM.

Would like to catch the timeout for image polling.

File: provider/image.go [hunk 6]
```diff
@@ -370,28 +358,62 @@ func (p *dockerNativeProvider) dockerBuild(ctx context.Context,
...
-	for pushScanner.Scan() {
-		info, err := processLogLine(pushScanner.Text())
-		if err != nil {
-			return "", nil, err
+	scanner := newScanner(pushOutput)
+	for scanner.Scan() {
+		line := scanner.Text()
+		if len(line) == bufio.MaxScanTokenSize {
+			_ = p.host.Log(ctx, "warning", urn,
+				fmt.Sprintf("Docker build output line exceeds %d KiB", bufio.MaxScanTokenSize/1024))
```
[the remark this answers is not in the document]
Comment (AaronFriel) [reply provider/image.go]: We can remove this, as we've proven that it happens in practice even if we can't reliably repro.

File: provider/image.go (hunk 3 above)
Comment (AaronFriel) [reply provider/image.go:520]: Yeah, we can cut this down to 5 seconds.

File: provider/image.go (hunk 4 above)
Comment (AaronFriel) [reply provider/image.go:531]: I think we should only log actionable messages to persistent output, avoid cluttering output.

File: provider/image.go (hunk 5 above)
Comment (AaronFriel) [reply provider/image.go:517]: Yeah, fixed by adding the error at the end of polling.

File: provider/image.go [hunk 7]
```diff
@@ -527,6 +533,9 @@ func (p *dockerNativeProvider) runImageBuild(
 		}
 		time.Sleep(1 * time.Second)
 	}
+	if imageID == "" {
```
Comment (guineveresaenger) [inline provider/image.go:536]: We have an `imageID == ""` on line 498 above, and not doing anything to it in between as far as I can tell.

Should a check be run in the final `findImageID` loop instead?

**Comment by guineveresaenger** [comment]
That looks great, thank you! Feel free to :shipit: when tests are green

**Comment by lionello** [comment]
I have the impression that this PR caused a regression in #32 because the random `build-id` appears to cause the image to change each time I run `pulumi up`.

**Pull request merged**